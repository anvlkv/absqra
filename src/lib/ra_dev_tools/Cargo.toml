[package]
name = "ra_dev_tools"
version = "0.1.0"
authors = ["a.nvlkv <a.nvlkv@protonmail.ch>"]
edition = "2018"
build = "build.rs"

[build-dependencies]
build-deps = "0.1.4"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = "0.16.0"
regex = "1.3.9"
example_files_macro = {path="./src/lib/example_files_macro"}
files_iterator = {path="./src/lib/files_iterator"}
