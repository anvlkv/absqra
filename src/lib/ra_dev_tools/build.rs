extern crate build_deps;

fn main() {
    // Enumerate files in sub-folder "examples/*", being relevant for the test-generation (as example)
    // If function returns with error, exit with error message.
    build_deps::rerun_if_changed_paths( "../../../examples/*/*" ).unwrap();

    // Adding the parent directory "examples" to the watch-list will capture new-files being added
    build_deps::rerun_if_changed_paths( "../../../examples/*" ).unwrap();
}