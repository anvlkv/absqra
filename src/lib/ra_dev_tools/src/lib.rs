pub extern crate insta;
pub use files_iterator::for_each_ra_example_file;

#[allow(unused_attributes)]
#[macro_use] pub use example_files_macro::make_example_tests;

#[cfg(test)]
mod test {
    use super::make_example_tests;

    #[make_example_tests]
    #[test]
    fn it_works_with(content: String, file_name: String) {
        println!("didn't panic with content: \n[{}] of file_name: {}", content, file_name);
    }
}