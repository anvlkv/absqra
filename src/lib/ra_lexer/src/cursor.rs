use std::str::Chars;
use std::fmt;
pub(crate) const EOF_CHAR: char = '\0';
pub(crate) const EOL_CHAR: char = '\n';

pub(crate) struct Cursor<'a> {
    initial_len: usize,
    chars: Chars<'a>,
    pub position: Position,
    pub level: usize
}

#[derive(Copy, Clone, PartialEq)]
pub struct Position(pub usize, pub usize);

impl fmt::Debug for Position {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "[{}:{}]", self.0, self.1)
    }
}



/// True if `c` is considered a whitespace
pub fn is_whitespace(c: char) -> bool {
    match c {
        // Usual ASCII suspects
        | '\u{0009}' // \t
        | '\u{000B}' // vertical tab
        | '\u{000C}' // form feed
        | '\u{0020}' // space

        // Bidi markers
        | '\u{200E}' // LEFT-TO-RIGHT MARK
        | '\u{200F}' // RIGHT-TO-LEFT MARK
            => true,
        _ => false,
    }
}

/// True if `c` is considered an end of line
pub fn is_end_of_line(c: char) -> bool {
    match c {
        // Usual ASCII suspects
        | '\u{000A}' // \n
        | '\u{000D}' // \r

        // NEXT LINE from latin1
        | '\u{0085}'

        // // Dedicated whitespace characters from Unicode
        | '\u{2028}' // LINE SEPARATOR
        | '\u{2029}' // PARAGRAPH SEPARATOR
            => true,
        _ => false,
    }
}

impl <'a> Cursor<'a> {
    pub(crate)fn new(input: &'a str, position: Position, level: usize) -> Cursor<'a> {
        Cursor {
            initial_len: input.len(),
            chars: input.chars(),
            position,
            level
        }
    }

    /// Returns nth character relative to the current cursor position.
    /// If requested position doesn't exist, `EOF_CHAR` is returned.
    /// However, getting `EOF_CHAR` doesn't always mean actual end of file,
    /// it should be checked with `is_eof` method.
    fn nth_char(&self, n: usize) -> char {
        match self.chars().nth(n).unwrap_or(EOF_CHAR) {
            c if is_end_of_line(c) => EOL_CHAR,
            t => t
        }
    }

    /// Returns a `Chars` iterator over the remaining characters.
    fn chars(&self) -> Chars<'a> {
        self.chars.clone()
    }

    /// Peeks the next symbol from the input stream without consuming it.
    pub(crate) fn first_ahead(&self) -> char {
        self.nth_char(0)
    }

    /// Peeks the second symbol from the input stream without consuming it.
    pub(crate) fn second_ahead(&self) -> char {
        self.nth_char(1)
    }

    /// Checks if there is nothing more to consume.
    pub(crate) fn is_eof(&self) -> bool {
        self.chars.as_str().is_empty()
    }

    /// Moves to the next character.
    pub(crate) fn bump(&mut self) -> Option<char> {
        let character = self.chars.next();
        match character {
            Some(ch) => {
                match ch {
                    c if is_end_of_line(c) => {
                        self.position.0 += 1;
                        self.consume_indent();
                        self.position.1 = self.level;
                        self.bump()
                    }
                    c => {
                        self.position.1 += 1;
                        Some(c)
                    }
                }
            },
            None => None
        }
        
    }

    /// Returns amount of already consumed symbols.
    pub(crate) fn len_consumed(&self) -> usize {
        self.initial_len - self.chars.as_str().len()
    }

    /// Consumes indent level and returns indentation level
    fn consume_indent(&mut self) {
        self.level = self.eat_while(|c| is_whitespace(c));
    }

    /// Eats symbols while predicate returns true or until the end of file is reached.
    /// Returns amount of eaten symbols.
    fn eat_while<F>(&mut self, mut predicate: F) -> usize
    where
        F: FnMut(char) -> bool,
    {
        let mut eaten: usize = 0;
        while predicate(self.first_ahead()) && !self.is_eof() {
            eaten += 1;
            self.bump();
        }

        eaten
    }
}

#[cfg(test)]
mod tests {
    use super::{Cursor, EOF_CHAR, Position};
    #[test]
    fn it_should_create() {
        Cursor::new("abc", Position(1, 0), 0);
    }

    #[test]
    fn it_should_give_first_char() {
        let cur = Cursor::new("abc", Position(1, 0), 0);
        assert_eq!(cur.first_ahead(), 'a');
    }

    #[test]
    fn it_should_give_second_char() {
        let cur = Cursor::new("abc", Position(1, 0), 0);
        assert_eq!(cur.second_ahead(), 'b');
    }

    #[test]
    fn it_should_return_next_char() {
        let mut cur = Cursor::new("a", Position(1, 0), 0);
        assert_eq!(cur.bump().unwrap(), 'a');
    }

    #[test]
    fn it_should_return_none_at_end_of_input() {
        let mut cur = Cursor::new("a", Position(1, 0), 0);
        cur.bump();
        assert_eq!(cur.bump(), None);
    }

    #[test]
    fn it_should_confirm_eof() {
        let mut cur = Cursor::new("a", Position(1, 0), 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.is_eof(), true);
    }

    #[test]
    fn it_should_return_amount_of_consumed_symbols() {
        let mut cur = Cursor::new("abc", Position(1, 0), 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.len_consumed(), 3);
    }

    #[test]
    fn it_should_not_panic_when_encoutering_funny_characters() {
        let mut cur = Cursor::new("ðŸš¬", Position(1, 0), 0);
        assert_eq!(cur.bump().unwrap(), 'ðŸš¬');
    }

    #[test]
    fn it_should_track_position() {
        let mut cur = Cursor::new("abc\nbca", Position(1, 0), 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.position, Position(1, 1));
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.position, Position(1, 2));
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.position, Position(1, 3));
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.position, Position(2, 1));
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.position, Position(2, 2));
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.position, Position(2, 3));
    }

    #[test]
    fn it_should_track_indent_level() {
        let mut cur = Cursor::new("a\n\tb\n\t\tc\nd", Position(1, 0), 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.level, 0);
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.level, 1);
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.level, 2);
        assert_eq!(cur.position, Position(3, 3));
        assert_eq!(cur.bump().unwrap(), 'd');
        assert_eq!(cur.level, 0);
    }
}