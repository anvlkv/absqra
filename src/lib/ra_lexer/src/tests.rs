use crate::cursor::{Cursor, Position};
use crate::{tokenize, RaToken, TokenKind};

mod cursor {
    use super::{Cursor, Position};
    #[test]
    fn it_should_create() {
        Cursor::new("abc", Position(1, 0), 1, 0);
    }

    #[test]
    fn it_should_give_first_char() {
        let cur = Cursor::new("abc", Position(1, 0), 1, 0);
        assert_eq!(cur.first_ahead(), 'a');
    }

    #[test]
    fn it_should_give_second_char() {
        let cur = Cursor::new("abc", Position(1, 0), 1, 0);
        assert_eq!(cur.second_ahead(), 'b');
    }

    #[test]
    fn it_should_return_next_char() {
        let mut cur = Cursor::new("a", Position(1, 0), 1, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
    }

    #[test]
    fn it_should_return_none_at_end_of_input() {
        let mut cur = Cursor::new("a", Position(1, 0), 1, 0);
        cur.bump();
        assert_eq!(cur.bump(), None);
    }

    #[test]
    fn it_should_confirm_eof() {
        let mut cur = Cursor::new("a", Position(1, 0), 1, 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.is_eof(), true);
    }

    #[test]
    fn it_should_return_amount_of_consumed_symbols() {
        let mut cur = Cursor::new("abc", Position(1, 0), 1, 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.len_consumed(), 3);
    }

    #[test]
    fn it_should_not_panic_when_encountering_funny_characters() {
        let mut cur = Cursor::new("ðŸš¬", Position(1, 0), 1, 0);
        assert_eq!(cur.bump().unwrap(), 'ðŸš¬');
    }

    #[test]
    fn it_should_track_position() {
        let mut cur = Cursor::new("abc\nSOME", Position(1, 0), 1, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.position, Position(1, 1));
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.position, Position(1, 2));
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.position, Position(1, 3));
        assert_eq!(cur.bump().unwrap(), 'S');
        assert_eq!(cur.position, Position(2, 1));
        assert_eq!(cur.bump().unwrap(), 'O');
        assert_eq!(cur.position, Position(2, 2));
        assert_eq!(cur.bump().unwrap(), 'M');
        assert_eq!(cur.position, Position(2, 3));
    }

    #[test]
    fn it_should_track_indent_level() {
        let mut cur = Cursor::new("a\n\tb\n\t\tc\nd", Position(1, 0), 1, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.level, 1);
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.level, 2);
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.level, 3);
        assert_eq!(cur.position, Position(3, 3));
        assert_eq!(cur.bump().unwrap(), 'd');
        assert_eq!(cur.level, 1);
    }
}

mod lib {
    use super::{tokenize, Position, RaToken, TokenKind};

    #[test]
    fn it_should_create_iterator_of_tokens() {
        let mut stream = tokenize("abc");
        stream.next();
    }

    #[test]
    fn it_should_parse_multiple_lines() {
        let mut stream = tokenize("!?&/\n\tabc");
        stream.next();
        stream.next();
        stream.next();
        stream.next();
        assert_eq!(
            stream.next().unwrap().unwrap(),
            RaToken {
                kind: TokenKind::Identifier("abc"),
                content: "abc",
                position: (Position(2, 1), Position(2, 4)),
                len: 3,
                level: 2,
            }
        )
    }

    #[test]
    fn it_should_parse_identifiers() {
        let mut stream = tokenize("abc");
        assert_eq!(
            stream.next().unwrap().unwrap(),
            RaToken {
                kind: TokenKind::Identifier("abc"),
                content: "abc",
                position: (Position(1, 0), Position(1, 3)),
                len: 3,
                level: 1
            }
        );
    }

    use ra_dev_tools::insta::assert_json_snapshot;
    use ra_dev_tools::make_example_tests;

    #[make_example_tests]
    #[test]
    fn it_should_match_snapshots(contents: String, file_name: String) {
        let tokens: Vec<RaToken> = tokenize(&contents).map(|r| r.unwrap()).collect();
        assert_json_snapshot!(file_name, tokens)
    }

    mod strings {
        use super::{tokenize, Position, RaToken, TokenKind};
        #[test]
        fn it_should_parse_string_literals_with_double_quotes() {
            let mut stream = tokenize("\"some\"");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::StringLiteral("some"),
                    content: "some",
                    len: 6,
                    position: (Position(1, 0), Position(1, 6)),
                    level: 1
                }
            )
        }

        #[test]
        fn it_should_parse_string_literals_with_single_quotes() {
            let mut stream = tokenize("\'some\'");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::StringLiteral("some"),
                    content: "some",
                    len: 6,
                    position: (Position(1, 0), Position(1, 6)),
                    level: 1
                }
            )
        }
    }
    mod comments {
        use super::{tokenize, Position, RaToken, TokenKind};
        #[test]
        fn it_should_parse_single_line_comments() {
            let mut stream = tokenize("//abc");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "abc",
                    position: (Position(1, 0), Position(1, 5)),
                    len: 5,
                    level: 1
                }
            );
        }
        #[test]
        fn it_should_parse_multi_line_comments() {
            let mut stream = tokenize("/*abc\nSOME*/");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "abc\nSOME",
                    position: (Position(1, 0), Position(2, 6)),
                    len: 12,
                    level: 1
                }
            );
        }

        #[test]
        fn it_should_parse_multi_line_comments_with_fringe() {
            let mut stream = tokenize("/*\n* abc\n* SOME\n*/");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "\n* abc\n* SOME\n",
                    position: (Position(1, 0), Position(4, 2)),
                    len: 18,
                    level: 1
                }
            );
        }

        #[test]
        fn it_should_parse_comments_followed_by_one_another() {
            let mut stream = tokenize("//some\n\n/*\n* abc\n* SOME\n*/");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "some",
                    position: (Position(1, 0), Position(1, 6)),
                    len: 6,
                    level: 1
                }
            );

            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "\n* abc\n* SOME\n",
                    position: (Position(3, 0), Position(6, 2)),
                    len: 18,
                    level: 1
                }
            );
        }
    }
    
    mod numbers {
        use super::{tokenize, Position, RaToken, TokenKind};
        use crate::errors::LexerError;
        #[test]
        fn it_should_parse_numbers() {
            let mut stream = tokenize("123");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Int(123),
                    content: "123",
                    position: (Position(1, 0), Position(1, 3)),
                    len: 3,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Int(v) => assert_eq!(v, 123),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_separator() {
            let mut stream = tokenize("123,321");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123.321),
                    content: "123,321",
                    position: (Position(1, 0), Position(1, 7)),
                    len: 7,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123.321),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_another_decimal_separator() {
            let mut stream = tokenize("123.321");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123.321),
                    content: "123.321",
                    position: (Position(1, 0), Position(1, 7)),
                    len: 7,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123.321),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_and_thousands_separator() {
            let mut stream = tokenize("123.321,456");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123321.456),
                    content: "123.321,456",
                    position: (Position(1, 0), Position(1, 11)),
                    len: 11,
                    level: 1
                }
            );

            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123321.456),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_and_multiple_thousands_separator() {
            let mut stream = tokenize("123.321.123,456");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123321123.456),
                    content: "123.321.123,456",
                    position: (Position(1, 0), Position(1, 15)),
                    len: 15,
                    level: 1
                }
            );

            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123321123.456),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_another_decimal_and_multiple_thousands_separators() {
            let mut stream = tokenize("123,321,123.456");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123321123.456),
                    content: "123,321,123.456",
                    position: (Position(1, 0), Position(1, 15)),
                    len: 15,
                    level: 1
                }
            );

            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123321123.456),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_error_when_encountering_multiple_decimal_separators() {
            let mut stream = tokenize("123.321.123,456,654");
            match stream.next().unwrap() {
                Ok(_) => panic!("didn't error"),
                Err(e) => assert!(match e {
                    LexerError::UnexpectedCharacter(ch, _) => ch == ',',
                    _ => false
                })
            }
        }

        #[test]
        fn it_should_parse_negative_integer() {
            let mut stream = tokenize("-123");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Int(-123),
                    content: "-123",
                    position: (Position(1, 0), Position(1, 4)),
                    len: 4,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Int(v) => assert_eq!(v, -123),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_negative_float() {
            let mut stream = tokenize("-123.312");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(-123.312),
                    content: "-123.312",
                    position: (Position(1, 0), Position(1, 8)),
                    len: 8,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, -123.312),
                _ => panic!("unexpected kind")
            };
        }
    }

    mod content_block {
        use super::{tokenize, Position, RaToken, TokenKind};
        use crate::errors::LexerError;

        #[test]
        fn it_should_parse_content_blocks() {
            let mut stream = tokenize("`abc`");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::ContentBlock,
                    content: "abc",
                    position: (Position(1, 0), Position(1, 5)),
                    len: 5,
                    level: 1
                }
            );
        }

        #[test]
        fn it_should_keep_inner_indents_when_parsing_content() {
            let mut stream = tokenize("`\n\tabc\n\t\tabc\n`");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::ContentBlock,
                    content: "\n\tabc\n\t\tabc\n",
                    position: (Position(1, 0), Position(4, 1)),
                    len: 14,
                    level: 1
                }
            );
        }

        #[test]
        fn it_should_error_when_content_block_is_not_closed_properly() {
            let mut stream = tokenize("`\n\tabc`");
            match stream.next().unwrap() {
                Ok(_) => panic!("parsed invalid content"),
                Err(e) => assert!(match e {
                    LexerError::UnexpectedEndOfInput(_) => true,
                    _ => false
                })
            }
        }
    }

    mod single_char_tokens {
        use super::{tokenize, Position, RaToken, TokenKind};
        #[test]
        fn it_should_parse_tokens() {
            let mut stream = tokenize("!?&/#");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Exclamation,
                    position: (Position(1, 0), Position(1, 1)),
                    content: "!",
                    level: 1,
                    len: 1
                }
            );
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Question,
                    position: (Position(1, 1), Position(1, 2)),
                    content: "?",
                    level: 1,
                    len: 1
                }
            );
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Ampersand,
                    position: (Position(1, 2), Position(1, 3)),
                    content: "&",
                    level: 1,
                    len: 1
                }
            );
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Slash,
                    position: (Position(1, 3), Position(1, 4)),
                    content: "/",
                    level: 1,
                    len: 1
                }
            );
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::HashPound,
                    position: (Position(1, 4), Position(1, 5)),
                    content: "#",
                    level: 1,
                    len: 1
                }
            );
        }
    }


}
