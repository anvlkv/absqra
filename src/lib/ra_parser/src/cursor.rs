use ra_lexer::token::{Token};
use ra_lexer::cursor::Position;
use core::iter::Peekable;

pub(crate) struct Cursor<'token, I> 
    where I: Iterator<Item = Token<'token>> 
{
    pub position: Position,
    pub level: u16,
    tokens: Peekable<I>,
    consumed_len: usize
}

impl <'token, I> Cursor<'token, I> 
where I: Iterator<Item = Token<'token>> 
{
    pub fn new(tokens: I) -> Self
        
    {
        Self {
            position: Position(0, 0),
            level: 0,
            tokens: tokens.peekable(),
            consumed_len: 0
        }
    }

    pub fn first_ahead(&mut self) -> Option<&Token<'token>> {
        self.tokens.peek()
    }

    pub fn bump(&mut self) -> Option<Token<'token>> {
        let tok = self.tokens.next();
        match tok {
            Some(token) => {
                self.position = token.position.1;
                self.level = token.level;
                self.consumed_len += 1;
                Some(token)
            }
            None => None
        }
    }

    pub fn is_eof(& mut self) -> bool {
        self.first_ahead() == None
    }

    // pub fn do_while<F, D>(&mut self, mut test: F, mut cb: D)
    //     where F: FnMut (u16, Position, TokenKind) -> bool,
    //     D: FnMut (Token)
    // {
    //     loop {
    //         let next_token = self.first_ahead();
    //         match next_token {
    //             Some(t) => {
    //                 if test(t.level, t.position.0, t.kind) {
    //                     cb(self.bump().unwrap())
    //                 }
    //                 else {
    //                     break
    //                 }
    //             },
    //             None => break
    //         }
    //     }
    // }

    // pub fn read_within(&mut self, open: TokenKind, close: TokenKind) -> Vec<Token> {
    //     let mut read: Vec<Token> = Vec::new();

    //     loop {
    //         let next_token = self.first_ahead();

    //         match next_token {
    //             Some(tok) => {
    //                 match tok.kind {
    //                     k if k == open => {
    //                         let open_token = self.bump().unwrap();
    //                         let mut nested = self.read_within(open_token.kind, close.clone());
    //                         read.append(&mut nested);
    //                     },
    //                     k if k == close => {
    //                         self.bump();
    //                         break
    //                     },
    //                     _ => read.push(self.bump().unwrap())
    //                 }
    //             },
    //             None => break
    //         }

    //         // println!("{:?}", &read);
    //     }

    //     read
    // }
}