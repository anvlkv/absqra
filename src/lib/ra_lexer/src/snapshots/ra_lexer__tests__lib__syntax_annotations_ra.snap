---
source: src/lib/ra_lexer/src/tests.rs
expression: tokens
---
[
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        1,
        0
      ],
      [
        1,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "annotation"
    },
    "len": 10,
    "content": "annotation",
    "position": [
      [
        1,
        1
      ],
      [
        1,
        11
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 21,
    "content": " this is annotation",
    "position": [
      [
        1,
        12
      ],
      [
        1,
        33
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        3,
        0
      ],
      [
        3,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "media_context"
    },
    "len": 13,
    "content": "media_context",
    "position": [
      [
        3,
        1
      ],
      [
        3,
        14
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        3,
        14
      ],
      [
        3,
        15
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "media_option"
    },
    "len": 12,
    "content": "media_option",
    "position": [
      [
        3,
        15
      ],
      [
        3,
        27
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 43,
    "content": " ...[:media_option] more options possible",
    "position": [
      [
        3,
        28
      ],
      [
        3,
        71
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        5,
        0
      ],
      [
        5,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "annotation"
    },
    "len": 10,
    "content": "annotation",
    "position": [
      [
        5,
        1
      ],
      [
        5,
        11
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        5,
        11
      ],
      [
        5,
        12
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "option_a"
    },
    "len": 8,
    "content": "option_a",
    "position": [
      [
        5,
        12
      ],
      [
        5,
        20
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        5,
        20
      ],
      [
        5,
        21
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "option_b"
    },
    "len": 8,
    "content": "option_b",
    "position": [
      [
        5,
        21
      ],
      [
        5,
        29
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        5,
        29
      ],
      [
        5,
        30
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "option_c"
    },
    "len": 8,
    "content": "option_c",
    "position": [
      [
        5,
        30
      ],
      [
        5,
        38
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        5,
        38
      ],
      [
        5,
        39
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "option_d"
    },
    "len": 8,
    "content": "option_d",
    "position": [
      [
        5,
        39
      ],
      [
        5,
        47
      ]
    ],
    "level": 1
  }
]
